sympy/plotting/experimental_lambdify.py

```diff
@@ -154,41 +159,41 @@ class lambdify(object):
     """
 
     def __init__(self, args, expr):
+        from sympy.utilities.lambdify import lambdify
         self.args = args
         self.expr = expr
-        self.lambda_func_1 = experimental_lambdify(
+
+        func_1 = lambdify(args, expr)
+        func_2 = lambdify(args, expr, modules='mpmath')
+        func_3 = experimental_lambdify(
             args, expr, use_python_cmath=True, use_evalf=True)
-        self.lambda_func_2 = experimental_lambdify(
+        func_4 = experimental_lambdify(
             args, expr, use_python_math=True, use_evalf=True)
-        self.lambda_func_3 = experimental_lambdify(
+        func_5 = experimental_lambdify(
             args, expr, use_evalf=True, complex_wrap_evalf=True)
-        self.lambda_func = self.lambda_func_1
-        self.failure = False
+
+        self.candidates = [func_1, func_2, func_3, func_4, func_5]
 
     def __call__(self, args):
         try:
             #The result can be sympy.Float. Hence wrap it with complex type.
-            result = complex(self.lambda_func(args))
+            result = self.candidates[0](args)
+            result = complex(result)
             if abs(result.imag) > 1e-7 * abs(result):
                 return None
             return result.real
-        except (ZeroDivisionError, TypeError) as e:
+        except (ZeroDivisionError, TypeError, NameError) as e:
```

Comment:
I don't like catching `NameError` in general. If it is necessary then the `try/except` should be as narrow as possible. It would be better to fix the source of the `NameError` and not catch it here if possible.

Reply:
Nameerror is a problem of lambdify because it tries to print out unknown functions with adhoc names without any warnings that it actually exists in numpy

Reply:
It is possible that lambdify could changed to automatically detect unknown functions during the printing step, however, it needs bigger changes for printer and lambdify

Reply:
Okay well we can at least narrow the scope of the try/except. It should only contain the call to the function that could possibly raise `NameError`. Most of the time `NameError` is a bug in the code so we should really avoid catching it as much as possible.

Reply:
As a workaround (instead of a deeper fix), you could pass a custom printer to lambdify. For example, with `numpy`:
```
class CustomNumpyPrinter(sympy.printing.pycode.NumPyPrinter):
        def _print_not_supported(self, expr): 
            raise NotImplementedError()

printer = CustomNumpyPrinter({'fully_qualified_modules': False, 'inline': True, 'allow_unknown_functions': False})
try:
    lambda_func = sympy.lambdify(args, expr, modules=["numpy"], printer=printer)
except NotImplementedError:
    # Fallback to other method
```
Or perhaps it makes sense for that behaviour to be part of `lambdify` itself?